### PR TITLE
Added tags for iam policy

### PR DESCRIPTION
**Pull request by ra-grover**
### Description of your changes

This PR adds feature to tag IamPolicy, fixes https://github.com/crossplane/provider-aws/issues/599

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
The changes have been tested on a local kind cluster and verified the state of IamPolicy via cli.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: package/crds/identity.aws.crossplane.io_iampolicies.yaml [hunk 1]
```diff
@@ -78,6 +78,22 @@ spec:
                   path:
                     description: The path to the policy.
                     type: string
+                  tags:
+                    items:
+                      description: Tag represents a tag attached to a v1alpha1.User
```
Comment (chlunde) [inline package/crds/identity.aws.crossplane.io_iampolicies.yaml:83]: `description: Tag represents a tag attached to a v1alpha1.User`. Something went wrong here due to sharing the struct. Maybe the Tag struct doc strings could be changed to be generic?

File: apis/identity/v1alpha1/iampolicy_types.go [hunk 2]
```diff
@@ -37,6 +37,8 @@ type IAMPolicyParameters struct {
 
 	// The name of the policy.
 	Name string `json:"name"`
+
+	Tags []Tag `json:"tags,omitempty"`
```
Comment (chlunde) [inline apis/identity/v1alpha1/iampolicy_types.go:41]: Most of the time there's a doc string. Not sure how helpful this is in this case, but still:

Maybe something like.
```
       // Tags attached to the policy. For more information about
       // tagging, see Tagging customer managed policies (https://docs.aws.amazon.com/IAM/latest/UserGuide/id_tags.html)
       // in the IAM User Guide.
```

File: package/crds/identity.aws.crossplane.io_iampolicies.yaml (hunk 1 above)
Comment (ra-grover) [reply package/crds/identity.aws.crossplane.io_iampolicies.yaml:83]: Sure @chlunde , I will make a change. I copied that bit from IAMRole crd.

File: apis/identity/v1alpha1/iampolicy_types.go (hunk 2 above)
Comment (ra-grover) [reply apis/identity/v1alpha1/iampolicy_types.go:41]: Sure thing.

Comment (ra-grover) [reply apis/identity/v1alpha1/iampolicy_types.go:46]: Added @chlunde , marking this as resolved.

File: pkg/clients/iam/iamrole.go [hunk 3]
```diff
@@ -198,25 +198,23 @@ func IsRoleUpToDate(in v1beta1.IAMRoleParameters, observed iamtypes.Role) (bool,
...
 
 // DiffIAMTags returns the lists of tags that need to be removed and added according
-// to current and desired states.
-func DiffIAMTags(local []v1beta1.Tag, remote []iamtypes.Tag) (add []iamtypes.Tag, remove []string) {
-	addMap := make(map[string]string, len(local))
-	for _, t := range local {
-		addMap[t.Key] = t.Value
-	}
+// to current and desired states, also returns if desired state needs to be updated
+func DiffIAMTags(local map[string]string, remote []iamtypes.Tag) (add []iamtypes.Tag, remove []string, areTagsUpToDate bool) {
```
Comment (chlunde) [inline pkg/clients/iam/iamrole.go:202]: Sorry I missed this. This was a pure function, and now mutates the input. I don't know if crossplane has any guidelines around this but I think it is less error prone. Have you checked all callers?

cc @haarchri what do you think?

Comment (ra-grover) [reply pkg/clients/iam/iamrole.go:202]: Hey @chlunde yeah I have checked all the callers. I modified this to actually reuse this piece. `IamRoles/IamPolicies` both are using this for diffing the tags.

@haarchri Sir, any inputs ?

File: package/crds/identity.aws.crossplane.io_iampolicies.yaml (hunk 1 above)
Comment (ra-grover) [reply package/crds/identity.aws.crossplane.io_iampolicies.yaml]: @chlunde  My bad, `make` is dynamically generating this description. And is consistent with `IAMRole` crd. Reverting to what was already there, as `make check-diff` is breaking.

File: pkg/clients/iam/iamrole.go (hunk 3 above)
Comment (haarchri) [reply pkg/clients/iam/iamrole.go:202]: @chlunde is this a Blocker for you for me test was successful and tests incode looking AS expected

Comment (chlunde) [reply pkg/clients/iam/iamrole.go:202]: @haarchri @ra-grover no, it's good. I see the reason is that local is not a map, it used to be a slice. All callers take a copy of the input before calling this function so there should not be any bugs.

I hope we can make a generic version of DiffTags when Go 1.18 is out.. :)

**Review by chlunde — APPROVED** [review]
Testet tagging/untagging/retagging policy and role - works :)

**Comment by chlunde** [comment]
Thanks, merging. The codecov report points seems to be incorrect, it reports changes in coverage in methods you have not changed.

**Pull request merged**